### PR TITLE
Support TLS sockets

### PR DESCRIPTION
**Pull request by ivarref**
Please see issue: https://github.com/nrepl/nrepl/issues/266

`make check` passes.

Thanks!

Before submitting a PR make sure the following things have been done:

- [X] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [X] You've added tests to cover your change(s)
- [X] All tests are passing
- [X] The new code is not generating reflection warnings
- [X] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

Thanks!

*If you're just starting out to hack on nREPL you might find this [section of its
manual][1] extremely useful.*

[1]: https:/nrepl.org/nrepl/hacking_on_nrepl.html

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 ## master (unreleased)
 
+## 1.0.1 (2022-09-??)
+
+### New Features
+
+* [#???](https://github.com/nrepl/nrepl/pull/???): Add TLS support.
```
Comment (bbatsov) [inline CHANGELOG.md:9]: I'd used the ticket number #266 here.

File: src/clojure/nrepl/tls_client_proxy.clj [hunk 2]
```diff
@@ -0,0 +1,235 @@
+(ns nrepl.tls-client-proxy
```
Comment (bbatsov) [inline src/clojure/nrepl/tls_client_proxy.clj:1]: This namespace should have some high-level description.

File: src/clojure/nrepl/tls_client_proxy.clj [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  (:import (java.net Socket ServerSocket)
+           (java.io IOException BufferedInputStream BufferedOutputStream Closeable InputStream OutputStream)
+           (java.time.format DateTimeFormatter)
+           (java.time ZonedDateTime)))
+
+(def ^DateTimeFormatter pattern (DateTimeFormatter/ofPattern "yyyy-MM-dd HH:mm:ss z"))
+
+(defonce lock (Object.))
+
+(defn atomic-println [& args]
```
Comment (bbatsov) [inline src/clojure/nrepl/tls_client_proxy.clj:14]: Please make sure that all public functions have docstrings. If something's not part of the public API then mark it as private explicitly.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,12 @@
 
 ## master (unreleased)
 
+## 1.0.1 (2022-09-??)
```
Comment (bbatsov) [inline CHANGELOG.md:5]: You can remove this line.

**Comment by bbatsov** [comment]
Overall the changes look good to me, apart from my small inline remarks.

File: doc/modules/ROOT/pages/usage/tls.adoc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+Sometimes however `ssh` may not be available, and thus you need to
+bind to `0.0.0.0` to make the nREPL server available. The following will explain
+how to do that with TLS.
+
+== Generating certificates and keys
+
+[source,shell]
+----
+$ clojure -Sdeps '{:aliases {:write-cert {:deps {com.github.ivarref/locksmith {:mvn/version "0.1.6"}} :exec-fn com.github.ivarref.locksmith/write-certs!}}}' -T:write-cert :duration-days 365
```
Comment (bbatsov) [inline doc/modules/ROOT/pages/usage/tls.adoc:14]: Might be good to also show some examples how to generate the certs with standard Unix utils (if possible).

File: doc/modules/ROOT/pages/usage/tls.adoc [hunk 6]
```diff
@@ -0,0 +1,56 @@
+= TLS (Transport Layer Security) support
+
+The standard way to securely access a nREPL server has been
+to use `ssh` with port forwarding, and letting the nREPL server only bind to `127.0.0.1`.
+
+Sometimes however `ssh` may not be available, and thus you need to
```
Comment (bbatsov) [inline doc/modules/ROOT/pages/usage/tls.adoc:6]: I'd expand a little bit on this section - e.g. mention that the connection over TLS will be encrypted and require a certain key to be established. We can't make too many assumptions about the background of the readers of the docs.

File: doc/modules/ROOT/pages/usage/tls.adoc [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                                 :tls? true ; will cause the server to fail if no certs/keys are given
+                                 ; if server.keys is on the server
+                                 :tls-keys-file "server.keys"))
+                                 ; use :tls-keys-str if the contents of server.keys is available as a string,
+                                 ; e.g.
+                                 ; :tls-keys-str (System/getenv "NREPL_SERVER_KEYS")
+='user/server
+----
+
+== Connect with cmdline client
```
Comment (bbatsov) [inline doc/modules/ROOT/pages/usage/tls.adoc:38]: Might be good to explain here what should 3rd party clients do to support TLS, at least at a high level.

**Comment by ivarref** [comment]
Hi again @bbatsov 

and thanks for the quick response.

I did some changes to what you suggested. What do you think?

Thanks and kind regards.

**Comment by bbatsov** [comment]
@ivarref The updates look good to. Thanks!

**Comment by bbatsov** [comment]
> This branch cannot be rebased due to conflicts

You might also want to rebase on top of `master` and squash the commits into one or several (e.g. 3-4).

**Comment by ivarref** [comment]
Thanks @bbatsov !

I've squashed it into a single commit now, adding an extra thank you to Aphyr, and setting a version number to 1.0.0.
What needs to be updated before a release is the version numbers used in the documentation (adoc and clj files).

Thanks and kind regards.

**Comment by ivarref** [comment]
I've squashed and pushed again now, fixing a codespell issue.

**Comment by bbatsov** [comment]
I was on vacation last week, but I'll aim to wrap this up in the upcoming week.

**Comment by bbatsov** [comment]
Something seems to be wrong with the CI, can you push again to force a re-run?

**Comment by ivarref** [comment]
I've force pushed again now.

**Comment by bbatsov** [comment]
Hmm, seems the CI is not working at all. I'm puzzled by the error message:

> Could not find a usable config.yml, you may have revoked the CircleCI OAuth app. 

`config.yml` is there and I haven't revoked the Circle's access to the repo.

**Comment by ivarref** [comment]
Have you tried re-authenticating? Could the access token be expired? 

Googling led me to this link:
https://support.circleci.com/hc/en-us/articles/360043002793-Troubleshooting-CircleCI-Access-After-Enabling-Github-SSO

**Comment by bbatsov** [comment]
Everything looks good in the admin panel. I might have to reach out to Circle CI's support for this.

**Comment by bbatsov** [comment]
Can you try again? I've just reset the CircleCI permissions on my account, which might help in theory.

**Comment by ivarref** [comment]
Another push, still looks to be failing :-/

**Comment by bbatsov** [comment]
That's the response I got from the support:

```
We recently made a change in how we fetch the pipeline configuration which explains the apparition of the outcome you've observed.Previously we would attempt to fetch the configuration using the token of one of the project's followers, as a fall back method.The change we implemented to ensure a more accurate error messaging, enforces the sole use of the pipeline triggerer's token. If that token is invalid we no longer fall back on another followers' token, and simply notify that we were not able to fetch the configuration.Could you please have the affected user log out of CircleCI, clear the cache and cookies of their browser and log back in? This would allow our system to reissue a token for your user, allowing them to trigger a build again. | We recently made a change in how we fetch the pipeline configuration which explains the apparition of the outcome you've observed.Previously we would attempt to fetch the configuration using the token of one of the project's followers, as a fall back method.The change we implemented to ensure a more accurate error messaging, enforces the sole use of the pipeline triggerer's token. If that token is invalid we no longer fall back on another followers' token, and simply notify that we were not able to fetch the configuration.Could you please have the affected user log out of CircleCI, clear the cache and cookies of their browser and log back in? This would allow our system to reissue a token for your user, allowing them to trigger a build again.

We recently made a change in how we fetch the pipeline configuration which explains the apparition of the outcome you've observed.Previously we would attempt to fetch the configuration using the token of one of the project's followers, as a fall back method.The change we implemented to ensure a more accurate error messaging, enforces the sole use of the pipeline triggerer's token. If that token is invalid we no longer fall back on another followers' token, and simply notify that we were not able to fetch the configuration.Could you please have the affected user log out of CircleCI, clear the cache and cookies of their browser and log back in? This would allow our system to reissue a token for your user, allowing them to trigger a build again.```

Can you log out of Circle CI to see if this will help?

**Comment by ivarref** [comment]
That worked! It's building now again. Let me fix the failing tests.

**Comment by ivarref** [comment]
All checks passed.

**Comment by bbatsov** [comment]
I'll cut a 1.1 alpha release soon, so people can test this more easily.

**Pull request merged**